File: 3742-maximum-path-score-in-a-grid/3742-maximum-path-score-in-a-grid.py
class Solution(object):
    def maxPathScore(self, grid, k):
        m, n = len(grid), len(grid[0])

        # dp[i][j][c] = max score
        dp = [[[-1] * (k + 1) for _ in range(n)] for _ in range(m)]

        dp[0][0][0] = 0

        for i in range(m):
            for j in range(n):
                for c in range(k + 1):
                    if dp[i][j][c] == -1:
                        continue

                    # move right
                    if j + 1 < n:
                        val = grid[i][j + 1]
                        cost = 0 if val == 0 else 1
                        score = val

                        if c + cost <= k:
                            dp[i][j + 1][c + cost] = max(
                                dp[i][j + 1][c + cost],
                                dp[i][j][c] + score
                            )

                    # move down
                    if i + 1 < m:
                        val = grid[i + 1][j]
                        cost = 0 if val == 0 else 1
                        score = val

                        if c + cost <= k:
                            dp[i + 1][j][c + cost] = max(
                                dp[i + 1][j][c + cost],
                                dp[i][j][c] + score
                            )

        # find best at destination
        ans = max(dp[m-1][n-1])

        return ans if ans != -1 else -1

File: 3742-maximum-path-score-in-a-grid/3742-maximum-path-score-in-a-grid.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number}
 */
var maxPathScore = function(grid, k) {
    const m = grid.length;
    const n = grid[0].length;

    // dp[i][j][c] = max score
    let dp = Array.from({ length: m }, () =>
        Array.from({ length: n }, () => Array(k + 1).fill(-1))
    );

    dp[0][0][0] = 0;

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            for (let c = 0; c <= k; c++) {
                if (dp[i][j][c] === -1) continue;

                // move right
                if (j + 1 < n) {
                    let val = grid[i][j + 1];
                    let cost = val === 0 ? 0 : 1;
                    let score = val;

                    if (c + cost <= k) {
                        dp[i][j + 1][c + cost] = Math.max(
                            dp[i][j + 1][c + cost],
                            dp[i][j][c] + score
                        );
                    }
                }

                // move down
                if (i + 1 < m) {
                    let val = grid[i + 1][j];
                    let cost = val === 0 ? 0 : 1;
                    let score = val;

                    if (c + cost <= k) {
                        dp[i + 1][j][c + cost] = Math.max(
                            dp[i + 1][j][c + cost],
                            dp[i][j][c] + score
                        );
                    }
                }
            }
        }
    }

    let ans = Math.max(...dp[m - 1][n - 1]);
    return ans === -1 ? -1 : ans;
};